addons/stock/models/stock_orderpoint.py

```diff
@@ -344,82 +344,53 @@ def _get_orderpoint_action(self):
         orderpoints_removed = orderpoints._unlink_processed_orderpoints()
         orderpoints = orderpoints - orderpoints_removed
         to_refill = defaultdict(float)
-        all_product_ids = []
-        all_warehouse_ids = []
-        # Take 3 months since it's the max for the forecast report
-        to_date = add(fields.date.today(), months=3)
-        qty_by_product_warehouse = self.env['report.stock.quantity']._read_group(
-            [('date', '=', to_date), ('state', '=', 'forecast')],
-            ['product_id', 'product_qty', 'warehouse_id'],
-            ['product_id', 'warehouse_id'], lazy=False)
-        for group in qty_by_product_warehouse:
-            warehouse_id = group.get('warehouse_id') and group['warehouse_id'][0]
-            if group['product_qty'] >= 0.0 or not warehouse_id:
-                continue
-            all_product_ids.append(group['product_id'][0])
-            all_warehouse_ids.append(warehouse_id)
-            to_refill[(group['product_id'][0], warehouse_id)] = group['product_qty']
-        if not to_refill:
-            return action
-
-        # Recompute the forecasted quantity for missing product today but at this time
-        # with their real lead days.
-        key_to_remove = []
-        pwh_per_day = defaultdict(list)
-        for (product, warehouse) in to_refill.keys():
-            product = self.env['product.product'].browse(product).with_prefetch(all_product_ids)
-            warehouse = self.env['stock.warehouse'].browse(warehouse).with_prefetch(all_warehouse_ids)
-            rules = product._get_rules_from_location(warehouse.lot_stock_id)
-            lead_days = rules.with_context(bypass_delay_description=True)._get_lead_days(product)[0]
-            pwh_per_day[(lead_days, warehouse)].append(product.id)
-        # group product by lead_days and warehouse in order to read virtual_available
+        all_product_ids = self.env['product.product'].search([('product_tmpl_id.type', '=', 'product')])
+        all_replenish_location_ids = self.env['stock.location'].search([('replenish_location', '=', True)])
+        ploc_per_day = defaultdict(list)
+        for product in all_product_ids:
+            for loc in all_replenish_location_ids:
+                rules = product._get_rules_from_location(loc)
+                lead_days = rules.with_context(bypass_delay_description=True)._get_lead_days(product)[0]
+                ploc_per_day[(lead_days, loc)].append(product.id)
```

Comment:
It will be take way too much time. It's true that it's will give a complete solution but in a first try I would only do a check on the forecast (without to_date specify). Then only compute with the lead time. It's true that we will lose this usecase:
- product with lead time 10 days.
- Stock at day 0: 0 units
- Delivery at day 10: 10 units
- Receipt in 3 months of 100 units
It doesn't trigger the rule to have the supply for the planned delivery

But they could create the orderpoint themself if it happens a lot for specific products